[pytorch] Minor: avoid acquiring GIL twice in PyRRef::localValue()

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32785 [pytorch] Minor: avoid acquiring GIL twice in PyRRef::localValue()**

Add PythonRpcHandler::handleExceptionWithGIL() so that in PyRRef::localValue(),
we don't need to release the GIL and re-acquire the following line.

Differential Revision: [D19626195](https://our.internmc.facebook.com/intern/diff/D19626195/)

Commit:
[pytorch] Minor: avoid acquiring GIL twice in PyRRef::localValue()

Add PythonRpcHandler::handleExceptionWithGIL() so that in PyRRef::localValue(),
we don't need to release the GIL and re-acquire the following line.

Differential Revision: [D19626195](https://our.internmc.facebook.com/intern/diff/D19626195/)

[ghstack-poisoned]

zhaojuanmao commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -61,13 +61,13 @@ py::object PyRRef::localValue() {
     // grabbing the GIL.
     pybind11::gil_scoped_acquire ag;
     res = torch::jit::toPyObject(std::move(value));
+    PythonRpcHandler::getInstance().handleExceptionGILHeld(res);
```
@jjlilley it could rarely happen for some cases, this thread one is waiting for PythonRpcHandler to be constructed by thread two,  thread two needs to acquire the GIL to construct the PythonRpcHandler, but thread two can not get the GIL, because thread one is holding it in this scope. In this rare case, dead lock could happen though

jjlilley replied:
Ah, thanks for the clarification. 
but it seems like I could safely acquire the PythonRpcHandler outside the lock, right?

  auto &rpcHandler = PythonRpcHandler::getInstance();
  pybind11::gil_scoped_acquire ag;
  res = torch::jit::toPyObject(std::move(value));
  rpcHandler.handleExceptionGILHeld(res);

zhaojuanmao replied:
yes, this works, thanks!

Commit:
Update on "[pytorch] Minor: avoid acquiring GIL twice in PyRRef::localValue()"


Add PythonRpcHandler::handleExceptionWithGIL() so that in PyRRef::localValue(),
we don't need to release the GIL and re-acquire the following line.

Differential Revision: [D19626195](https://our.internmc.facebook.com/intern/diff/D19626195/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 47957a54:


**None of the build failures appear to be your fault**.

* **2/2** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=50d82f51223726725317919d429a02979d9bc39c&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base 50d82f51 since Jan 29
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=47957a5444de191721d5a5007bd7781845413682).

### :construction: 2 upstream failures recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [caffe2_onnx_py3_6_clang7_ubuntu16_04_test](https://dr.pytorch.org/build-details.html?build_id=92775419) from Jan 29
* [caffe2_onnx_py2_gcc5_ubuntu16_04_test](https://dr.pytorch.org/build-details.html?build_id=92775510) from Jan 29



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
rebase on "[pytorch] Minor: avoid acquiring GIL twice in PyRRef::localValue()"


Add PythonRpcHandler::handleExceptionWithGIL() so that in PyRRef::localValue(),
we don't need to release the GIL and re-acquire the following line.

Differential Revision: [D19626195](https://our.internmc.facebook.com/intern/diff/D19626195/)

[ghstack-poisoned]

xush6528 reviewed (approved):
looks good
